pre-docs/writing_yaml_plans.md

```diff
@@ -473,3 +473,134 @@ plan mymodule::yamlplan(
+        content: $facts['os']['family']
+        ensure: present
+  - name: file_contents
+    description: Read contents of file managed with file resource
+    eval: >
+      file::read('/tmp/foo')
+      
+return: $file_contents
+```
+This plan performs apply prep on a localhost target, next it uses a Puppet `file` resource to write the OS family discovered from the Puppet `$facts` hash to a temporary file. Finally it reads the value written to the file and returns it. Running `bolt plan convert` on this plan produces the following puppet code:
+```
+plan yaml_plans::interpolation_pp() {
+  apply_prep('localhost')
+
+  $interpolation = apply('localhost') {
+    file { '/tmp/foo':
+      content => $facts['os']['family'],
+      ensure => 'present',
+    }
+  }
+  $file_contents = file::read('/tmp/foo')
+
+  return $file_contents
+}
+```
+In this case the Puppet plan produced from the Yaml plan works as expected where the Yaml version fails. The cause of the failure is that the `$facts` variable is attempted to be resolved as a plan variable instead of being evaluated as part of compiling the manifest code in an apply block.
+
+The next example also has a somewhat related interpolation issue. 
+
+## Dependency order
+
+The resources in a resource list are applied in order. It is still possible to explicitly set dependencies but it is important to note how to reference them. Consider the following Yaml plan:
+```
+parameters:
+  nodes:
+    type: TargetSpec
+steps:
+  - eval: |
+      apply_prep($nodes)
+  - name: pkg
+    target: $nodes
+    resources:
+      - title: openssh-server
+        type: package
+        parameters:
+          ensure: present
+          before: File['/etc/ssh/sshd_config']
+      - title: /etc/ssh/sshd_config
+        type: file
+        parameters:
+          ensure: file
+          mode: '0600'
+          content: ''
+          require: Package['openssh-server']
+```
+Executing this plan will fail during catalog compilation because of how the resources referenced in the `before` and `require` parameters are parsed. Specifically the error will be at `Could not find resource 'File['/etc/ssh/sshd_config']' in parameter 'before'`. The solution for this is to not quote the resource titles. See the corrected plan example below.
+```
+parameters:
+  nodes:
+    type: TargetSpec
+steps:
+  - eval: |
+      apply_prep($nodes)
+  - name: pkg
+    target: $nodes
+    resources:
+      - title: openssh-server
+        type: package
+        parameters:
+          ensure: present
+          before: File[/etc/ssh/sshd_config]
+      - title: /etc/ssh/sshd_config
+        type: file
+        parameters:
+          ensure: file
+          mode: '0600'
+          content: ''
+          require: Package[openssh-server]
+```
+It should be noted that in general resources should be declared in order and this particular use case is uncommon. It is meant to illustrate a case where parameter parsing can lead to some non-intuitive results.
```

Comment:
Recommend shortening:

>Note that in general, resources should be declared in order. This example is unusual because it is meant to illustrate a case where parameter parsing can lead to non-intuitive results.